### PR TITLE
Groovy 4 safe changes

### PR DESCRIPTION
**Pull request by DPUkyle**
These changes were uncovered by work on #20038, but are not reliant on Groovy 4. These should be safe to proactively merge and will reduce the friction of upgrading to Groovy 4.

### CODE REVIEW

**Comment by DPUkyle** [comment]
@bot-gradle test QF

**Comment by DPUkyle** [comment]
@bot-gradle test RFM

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/OutputDirSelectorUtil.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.performance.results;
+
+public class OutputDirSelectorUtil {
```
Comment (tresat) [inline subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/OutputDirSelectorUtil.java:19]: Javadoc on top-level type?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy [hunk 2]
```diff
@@ -374,4 +374,19 @@ resolve.doLast { assert ruleInvoked }
...
+        "(new javax.xml.namespace.QName('http://my.extra.info/${name}', '${name}'))"
+    }
+
+    private static String sq(String input) {
+        escapeForSingleQuoting(input)
+    }
+
+    private static String escapeForSingleQuoting(String input) {
+        input.replace('\\', '\\\\').replace('\'', '\\\'')
+    }
```
Comment (tresat) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy:391]: Maybe just `sq` is needed here.

**Review by tresat — APPROVED** [review]

File: subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/OutputDirSelectorUtil.java (hunk 1 above)
Comment (DPUkyle) [reply subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/OutputDirSelectorUtil.java:23]: I addressed this; thank you!

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy (hunk 2 above)
Comment (DPUkyle) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/IvySpecificComponentMetadataRulesIntegrationTest.groovy]: I addressed this; thank you!

File: subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
 import java.util.Iterator;
 import java.util.List;
 
+@SuppressWarnings("unchecked")
```
[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: @mlopatkin it looks like you have this change in your branch `ml/wip/groovy-indy-instrumentation-poc` - do you have any recollection about why this change is needed? It seems unnecessary. @lptr does ff108e8f115d090f12fb57033127f9688c76d8df ring a bell?

[the remark this answers is not in the document]
Comment (mlopatkin) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: I don't recall touching this file or seeing warnings in it. My branch was based on this one, that's how the change ended up there, I think. My guess is that this line is a leftover after a rebase/merge of the proper fix in 0dd87a0d62e2098c245a453a99b358cc4903aac1

[the remark this answers is not in the document]
Comment (DPUkyle) [reply subprojects/model-core/src/testFixtures/groovy/org/gradle/api/internal/provider/ProviderTestUtil.java:27]: 💯 Thanks Mike. I agree, the use of `Cast.uncheckedNonnullCast` removes the need for the annotation. I'll remove this.

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Review by nathan-contino — APPROVED** [review]
LGTM

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Pull request merged**